### DISCUSSION TITLE
Unknown df in get_farm_AEP_wind_rose_class

### DISCUSSION THREAD

**Discussion by JarneVB** [NatLabRockies/floris#529] [Q&A] [answered]
I tried calculating the farm AEP with a wind rose by using get_farm_AEP_wind_rose_class, but I get the following error message: 
> 'DataFrame' object has no attribute 'df'

It is about these lines:

       wind_speeds_wind_rose = wind_rose.df.ws.unique()
       wind_directions_wind_rose = wind_rose.df.wd.unique()
       self.reinitialize(wind_speeds=wind_speeds_wind_rose, wind_directions=wind_directions_wind_rose)
       # Build the frequency matrix from wind rose
       freq = wind_rose.df.set_index(['wd','ws']).unstack().values`

It looks like df is defined nowhere. How can I fix this?

**Comment by paulf81** [comment]
hi @JarneVB, I would like to help.  Just to check something quickly, do you get this error if you run the example script: 08_calc_aep_from_rose_use_class.py ?

**Reply by JarneVB** [reply]
Thank you for your fast response! The example script works fine, it's just my own script. Might it have something to do with the wind rose argument that I use as input? I used a wind rose generated by make_wind_rose_from_weibull().

**Comment by paulf81** [comment]
Would it be possible for you to post a minimal version of the script that produces the error?  I'd be happy to debug

**Reply by JarneVB** [reply]
from floris.floris.tools.floris_interface import FlorisInterface 
    from floris.floris.tools import FlorisInterface, wind_rose, WindRose
    D = 112
    X_farm = [1*D, 10*D, 20*D]
    Y_farm = [1*D, 10*D, 20*D]
    
    fi_weibull_jensen = FlorisInterface("floris/examples/inputs/jensen.yaml")
    fi_weibull_jensen.reinitialize(layout_x=X_farm, layout_y=Y_farm)
    wind_data_weibull =  make_wind_rose_from_weibull()
    jensen_aep = fi_weibull_jensen.get_farm_AEP_wind_rose_class(wind_rose=wind_data_weibull)
    print("Farm AEP (default options): {:.3f} GWh".format(jensen_aep / 1.0e9))
    

I reinitialize for a farm of 50 turbines, but it also gives the same problem for less turbines, so this is a bit shorter.

**Reply by paulf81** [reply, marked as answer]
Hi!  Ok, I think there is an issue in our code, the function make_wind_rose_from_weibull() both returns a dataframe, and updates the member dataframe of the WindRose class.  We have a TODO item to remove the redundant return which I'll convert into an issue now.  In the meanwhile, this code works for me (ignoring the returned df and using the updated wind_rose object)

```
from floris.tools.floris_interface import FlorisInterface 
from floris.tools import FlorisInterface, WindRose

D = 112
X_farm = [1*D, 10*D, 20*D]
Y_farm = [1*D, 10*D, 20*D]

fi_weibull_jensen = FlorisInterface("inputs/jensen.yaml")
fi_weibull_jensen.reinitialize(layout_x=X_farm, layout_y=Y_farm)

wind_rose = WindRose()
_ =  wind_rose.make_wind_rose_from_weibull()

jensen_aep = fi_weibull_jensen.get_farm_AEP_wind_rose_class(wind_rose=wind_rose)
print("Farm AEP (default options): {:.3f} GWh".format(jensen_aep / 1.0e9))
```

**Reply by JarneVB** [reply]
This works!! Thank you for the help!

**Reply by JarneVB** [reply]
One more question: get_farm_AEP_wind_rose_class doesn't have an argument 'turbine_weights'. Is there another option to use the wake of neighboring wind farms, but excluding them from the AEP calculation?

**Reply by paulf81** [reply]
Yes, I think this just needs to turbine_weights added to the function get_farm_AEP_wind_rose_class, I'll open an issue

**Reply by paulf81** [reply]
Ok I've opened issue #540

**Reply by paulf81** [reply]
Now pull request #541

**Reply by JarneVB** [reply]
Wow thank you! This is very helpfull!

**Reply by paulf81** [reply]
No problem!
